python/cudf_polars/cudf_polars/experimental/benchmarks/pdsds_queries/q9.py

```diff
@@ -0,0 +1,153 @@
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 1 AND 20)
+           END bucket1,
+           CASE
+             WHEN (SELECT Count(*)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 21 AND 40) > 3392 THEN
+             (SELECT Avg(ss_ext_list_price)
+              FROM   store_sales
+              WHERE
+             ss_quantity BETWEEN 21 AND 40)
+             ELSE (SELECT Avg(ss_net_profit)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 21 AND 40)
+           END bucket2,
+           CASE
+             WHEN (SELECT Count(*)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 41 AND 60) > 32784 THEN
+             (SELECT Avg(ss_ext_list_price)
+              FROM   store_sales
+              WHERE
+             ss_quantity BETWEEN 41 AND 60)
+             ELSE (SELECT Avg(ss_net_profit)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 41 AND 60)
+           END bucket3,
+           CASE
+             WHEN (SELECT Count(*)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 61 AND 80) > 26032 THEN
+             (SELECT Avg(ss_ext_list_price)
+              FROM   store_sales
+              WHERE
+             ss_quantity BETWEEN 61 AND 80)
+             ELSE (SELECT Avg(ss_net_profit)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 61 AND 80)
+           END bucket4,
+           CASE
+             WHEN (SELECT Count(*)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 81 AND 100) > 23982 THEN
+             (SELECT Avg(ss_ext_list_price)
+              FROM   store_sales
+              WHERE
+             ss_quantity BETWEEN 81 AND 100)
+             ELSE (SELECT Avg(ss_net_profit)
+                   FROM   store_sales
+                   WHERE  ss_quantity BETWEEN 81 AND 100)
+           END bucket5
+    FROM   reason
+    WHERE  r_reason_sk = 1;
+    """
+
+
+def polars_impl(run_config: RunConfig) -> pl.LazyFrame:
+    """Query 9."""
+    # Load required tables
+    store_sales = get_data(run_config.dataset_path, "store_sales", run_config.suffix)
+    reason = get_data(run_config.dataset_path, "reason", run_config.suffix)
+
+    # Define bucket configurations: (min_qty, max_qty, count_threshold)
+    buckets = [
+        (1, 20, 3672),
+        (21, 40, 3392),
+        (41, 60, 32784),
+        (61, 80, 26032),
+        (81, 100, 23982),
+    ]
+
+    # Calculate each bucket value
+    bucket_values = []
+
+    for _, (min_qty, max_qty, threshold) in enumerate(buckets, 1):
+        # Count records in this quantity range
+        count = (
+            store_sales.filter(
+                pl.col("ss_quantity").is_between(min_qty, max_qty, closed="both")
+            )
+            .select(pl.len())
+            .collect()
```

Comment:
TODO: We're calling `collect` inside the query to produce an intermediary result in-memory. We should not do this. The entire query should be able to run on the GPU.

Reply:
Done in https://github.com/rapidsai/cudf/pull/19488/commits/66a770f5fa30dedef77e1de3a6885436267025c1